keps/sig-network/20200617-NodeIPAM-controller-support-variable-sized-CIDRs/README.md

```diff
@@ -0,0 +1,645 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Some user want to change `--node-cidr-mask-size` after create kubernetes cluster by some reason, 
+but current kubernetes doesn't allow `--node-cidr-mask-size` to change, because NodeIPAM controller
+doesn't support variable sized CIDRs.
+
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+Make NodeIPAM controller support variable sized CIDRs.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Use interval tree instead of bitmap to allocate podCIDR.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+For example, we have a kubernetes cluster, the `clusterCIDR` is 192.168.0.0/16 and `--node-cidr-mask-size` is 24
+1. add two nodes to cluster, podCIDR will be `192.168.0.0/24` and `192.168.1.0/24`;
+2. change `--node-cidr-mask-size` to 23 and delete node `192.168.1.0/24`;
+3. add a node pod CIDR will be `192.168.0.0/23`, `192.168.0.0/24` and `192.168.0.0/23` overlap.
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We need use a new data structure instead of bitmap to allocate podCIDR, it should support variable sized CIDRs, 
+interval tree is suitable. I found etcd already implement an interval tree 
+(https://github.com/etcd-io/etcd/blob/master/pkg/adt/interval_tree.go).
```

Comment:
Are there good benchmarks that show the memory footprint of such a thing?  I'd like to see some evidence that using it to store thousands of CIDRs uses reasonable amounts of memory.

Reply:
![profile001](https://user-images.githubusercontent.com/9653438/86095154-1a9c9280-bae4-11ea-8a01-710b5808a1ae.png)
I have add benchmarks.

Reply:
I think that it would be better to compare the results of the benchmarks between the old and the new implementation, ie. with 
https://pkg.go.dev/golang.org/x/tools/cmd/benchcmp

Reply:
![profile002](https://user-images.githubusercontent.com/9653438/86194953-630a8d80-bb82-11ea-86ce-2c1006794a7d.png)